lib/AST/NameLookup.cpp

```diff
@@ -1499,9 +1499,13 @@ bool DeclContext::lookupQualified(Type type,
     if (tracker)
       tracker->addUsedMember({current, member.getBaseName()},isLookupCascading);
 
-    // Make sure we've resolved implicit constructors, if we need them.
-    if (member.getBaseName() == ctx.Id_init && typeResolver)
-      typeResolver->resolveImplicitConstructors(current);
+    // Make sure we've resolved implicit members, if we need them.
+    if (typeResolver) {
+      if (member.getBaseName() == ctx.Id_init)
+        typeResolver->resolveImplicitConstructors(current);
+
+      typeResolver->resolveImplicitMember(current, member);
```

Comment:
@jrose-apple Following the protocol conformance checking is a bit hairy, but it looks like we should be in the clear. Once we've constructed the initial AST and have moved into type checking, we do breadth-first validation of members (at least, it's breadth-first for the first two passes), so types begin validating before references to these members can happen.

For a type like

```swift
struct OuterStruct {
    init() {
        innerStruct = InnerStruct.init(from: SomeDecoder(...))
    }

    let innerStruct: InnerStruct
    struct InnerStruct : Codable { let value: Int }
}
```
we first start validating `OuterStruct` and go member by member. Even though we hit `init()` first, we don't validate the body on the first pass, only that it has an interface type or that it's being validated. We hit `InnerStruct` and start validating protocol conformances before we hit the body of `init` and look up `init(from:)`.

By the time we ever end up doing a qualified lookup for `InnerStruct.init(from:)`, we've started checking the protocol conformance of `InnerStruct` to `Decodable`, so the state of the conformance is `Checking`. Once the conformance state is not `Incomplete`, we stop doing any lookups during checking.

**tl;dr:** We break recursion one the first pass, so this isn't an issue AFAICT.